### PR TITLE
[CI] Update MacOS version, move to Clang++ from ICPC

### PR DESCRIPTION
**Pull request by dmitriy-sobolev**
MacOS 11 was removed from GitHub CI on June 28 according to https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners. It has been replaced with MacOS 13.

Intel® C++ Compiler Classic Release was discontinued in the oneAPI 2024.0 release. It's been replaced with Clang++, which tends to be a default c++ compiler on MacOS.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -373,31 +371,23 @@ jobs:
...
+          - os: macos-13
+            cxx_compiler: clang++
             std: 17
             build_type: release
             backend: omp
     steps:
       - uses: actions/checkout@v3
-      - name: Install Intel® oneAPI C++ Compiler Classic and Intel® oneAPI Threading Building Blocks
+      - name: Install OpenMP for Clang++
+        if: (matrix.backend == 'omp' && matrix.cxx_compiler == 'clang++')
```
Comment (timmiesmith) [inline .github/workflows/ci.yml:382]: Is the if statement here necessary given that it exactly matches the matrix defined on the lines above?

**Review by timmiesmith — DISMISSED** [review]
LGTM. If the if check in the Install OpenMP isn't needed I can reapprove when you remove it.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (dmitriy-sobolev) [reply .github/workflows/ci.yml]: Done

**Review by timmiesmith — APPROVED** [review]
Looks good to me. Thank you.

**Pull request merged**